utils/binstat/binstat.go

```diff
@@ -0,0 +1,613 @@
+func               EnterTimeVal(what string, callerSize   int64 ) *BinStat      { return enterGeneric(what, true        , callerSize, sizeAtEnter, true ) }
+func               enterTimeVal(what string, callerSize   int64 ) *BinStat      { return enterGeneric(what, true        , callerSize, sizeAtEnter, false) }
+
+func (bs *BinStat) DebugParams (             callerParams string) *BinStat      {        bs.callerParams  = callerParams; return bs                       }
+func (bs *BinStat) Debug       (             text         string)               {        debugGeneric(bs  , text                                 , true ) }
+func (bs *BinStat) Run         (             f            func())               {        f           (                                                  ) }
+
+func (bs *BinStat) Point       (             pointUnique  string) time.Duration { return pointGeneric(bs  , pointUnique , 0         , sizeNotUsed, true ) }
+func (bs *BinStat) point       (             pointUnique  string) time.Duration { return pointGeneric(bs  , pointUnique , 0         , sizeNotUsed, false) }
+func (bs *BinStat) Leave       (                                ) time.Duration { return pointGeneric(bs  , "Leave"     , 0         , sizeNotUsed, true ) }
+func (bs *BinStat) leave       (                                ) time.Duration { return pointGeneric(bs  , "Leave"     , 0         , sizeNotUsed, false) }
+func (bs *BinStat) LeaveVal    (             callerSize   int64 ) time.Duration { return pointGeneric(bs  , "Leave"     , callerSize, sizeAtLeave, true ) }
+func (bs *BinStat) leaveVal    (             callerSize   int64 ) time.Duration { return pointGeneric(bs  , "Leave"     , callerSize, sizeAtLeave, false) }
+*/
+
+// functions W/&W/O go fmt ~ ~ ~ ~ ~ ~ ~ ~ ~ ~ ~ ~ ~ ~ ~ ~ ~ ~ ~ ~ ~ ~ ~ ~ ~ ~ ~ ~ ~ ~ ~ ~ ~ ~ ~ ~ ~ ~ ~ ~ ~ ~ ~ ~ ~ ~ ~ ~ ~
+
+func Enter(what string) *BinStat { return enterGeneric(what, false, 0, sizeNotUsed, true) }
+func enter(what string) *BinStat { return enterGeneric(what, false, 0, sizeNotUsed, false) }
+func EnterVal(what string, callerSize int64) *BinStat {
+	return enterGeneric(what, false, callerSize, sizeAtEnter, true)
+}
+func enterVal(what string, callerSize int64) *BinStat {
+	return enterGeneric(what, false, callerSize, sizeAtEnter, false)
+}
+func EnterTime(what string) *BinStat { return enterGeneric(what, true, 0, sizeNotUsed, true) }
+func enterTime(what string) *BinStat { return enterGeneric(what, true, 0, sizeNotUsed, false) }
+func EnterTimeVal(what string, callerSize int64) *BinStat {
+	return enterGeneric(what, true, callerSize, sizeAtEnter, true)
+}
+func enterTimeVal(what string, callerSize int64) *BinStat {
+	return enterGeneric(what, true, callerSize, sizeAtEnter, false)
+}
+
+func (bs *BinStat) DebugParams(callerParams string) *BinStat {
+	bs.callerParams = callerParams
+	return bs
+}
+func (bs *BinStat) Debug(text string) { debugGeneric(bs, text, true) }
+func (bs *BinStat) Run(f func())      { f() }
```

Comment:
Please note that unless we are going to encapsulate logic inside `Run`, this implementation is **redundant** and better to revert back. 

For example, assume the following code snippet that I borrowed [from your branch](https://github.com/onflow/flow-go/pull/891/files#diff-5e1fe1a3c4d732332e31ddbf630da35c7e8c9de10029f4f3f627045770cee32aR208-R211):
```
	bs := binstat.EnterTime(fmt.Sprintf("%s%s:%d", via, what, code))
	bs.Run(func() {
		data, err3 = cborEncMode.Marshal(v)
	})
	bs.LeaveVal(int64(len(data)))
```

This is _completely identical_ to the following version without adding any more value: 
```
	bs := binstat.EnterTime(fmt.Sprintf("%s%s:%d", via, what, code))
	data, err3 = cborEncMode.Marshal(v)
	bs.LeaveVal(int64(len(data)))
```

Now what I originally meant by [my comment](https://github.com/onflow/flow-go/pull/891#discussion_r670701993) was to encapsulate the `EnterTime` and `LeaveVal` logic into `Run` so that **invoking `Run` would implicitly handle the `binstat` bookkeeping**. Doing this way had significant advantages that we would not need to interleave same logic of entering and leaving all over the code, the code would look cleaner, and more maintainable over time, and more importantly, we would encapsulate and hide the underlying logic of `binstat` from the : 
```
func (bs *BinStat) RunWithLeaveVal(what string, val int64, f func()){
     enterTimeVal(what string)
     defer leaveVal(val)
     
    f()
}
```
Using the pattern above on the original example I made would look like: 
```
bs.Run(fmt.Sprintf("%s%s:%d", via, what, code), int64(len(data)), func() {
   data, err3 = cborEncMode.Marshal(v)
})
```
Note that my code may not be fully compilable, but I hope it can convey the gist. 

By reading your `binstat.go` I think in order to apply this comment you may need to have 2-3 versions of `Run`, e.g., `RunWithLeaveVal` (as I suggested above), `RunWithLeave` as I [suggested earlier](https://github.com/onflow/flow-go/pull/891#discussion_r670701993), etc. But it would result in a significantly cleaner, more readable, and easier-to-follow code. Though this is my personal opinion, so please look at it as a suggestion than a change request. 

Just please note that if you are not willing to follow this approach, I would ask you please remove `Run` entirely, as it does not have any extra value than inline invocation which was originally there.

Reply:
@yhassanzadeh13 thanks for the comment!

> Please note that unless we are going to encapsulate logic inside Run, this implementation is redundant and better to revert back.

I reverted!

> This is completely identical to the following version without adding any more value

Ah, sorry. That was a misunderstanding. I thought you originally wanted the anonymous function to make it more obvious which code is being monitored, and not necessarily to hide the `binstat` bookkeeping. That's the reason I added it even though it has no functional value apart from making things more obvious.

However, I think the code is identical in functionality but not in performance. According to a quick demo program I wrote then there is a wall-clock run-time overhead associated with calling anonymous functions, over doing the same thing without using the anonymous function. It's a similar deal and associated overhead of calling `binstat` and its function returning.

For these reasons, I decided to go down the commenting out the instrumentation by default route and according the last TPM meeting we can make an exception with `binstat` regarding commenting source code.

I like your idea to hide the bookkeeping AND have it as commented code by default. And maybe this can be done in a future PR with a future merciless refactor? François is also keen on having the auto commenting / uncommenting tool implemented as a Golang compiler plug-in, but I would also prefer to attack that in a future PR with a future merciless refactor too, since it could well be a big ticket item. Indeed, the two requests could end up in theory making each other more complicated to implement?

As such, and because `binstat` is still in its infancy and because this PR has taken longer already (although proven itself to be a wonderful Golang learning tool for me!) then I'm hoping to be pragmatic and hopefully move forward in the most agile way possible, while not closing the door to mercilessly refactoring `binstat` in a future PR request. I hope this makes sense! :-)